buffer/src/main/java/io/netty/buffer/api/CompositeBuffer.java

```diff
@@ -0,0 +1,2004 @@
+     */
+    @SafeVarargs
+    public static CompositeBuffer compose(BufferAllocator allocator, Send<Buffer>... sends) {
+        Buffer[] bufs = new Buffer[sends.length];
+        IllegalStateException ise = null;
+        for (int i = 0; i < sends.length; i++) {
+            if (ise != null) {
+                sends[i].close();
+            } else {
+                try {
+                    bufs[i] = sends[i].receive();
+                } catch (IllegalStateException e) {
+                    ise = e;
+                    for (int j = 0; j < i; j++) {
+                        bufs[j].close();
+                    }
+                }
+            }
+        }
+        if (ise != null) {
+            throw ise;
+        }
+        return new CompositeBuffer(allocator, filterExternalBufs(Arrays.stream(bufs)), COMPOSITE_DROP);
+    }
+
+    /**
+     * Create an empty composite buffer, that has no components. The buffer can be extended with components using either
+     * {@link #ensureWritable(int)} or {@link #extendWith(Send)}.
+     *
+     * @param allocator The allocator for the composite buffer. This allocator will be used e.g. to service
+     * {@link #ensureWritable(int)} calls.
+     * @return A composite buffer that has no components, and has a capacity of zero.
+     */
+    public static CompositeBuffer compose(BufferAllocator allocator) {
+        return new CompositeBuffer(allocator, EMPTY_BUFFER_ARRAY, COMPOSITE_DROP);
+    }
+
+    /**
+     * Check if the given buffer is a {@linkplain #compose(BufferAllocator, Send...) composite} buffer or not.
+     * @param composite The buffer to check.
+     * @return {@code true} if the given buffer was created with {@link #compose(BufferAllocator, Send...)},
+     * {@code false} otherwise.
+     */
+    public static boolean isComposite(Buffer composite) {
+        return composite.getClass() == CompositeBuffer.class;
+    }
+
+    private static Buffer[] filterExternalBufs(Stream<Buffer> refs) {
+        // We filter out all zero-capacity buffers because they wouldn't contribute to the composite buffer anyway,
+        // and also, by ensuring that all constituent buffers contribute to the size of the composite buffer,
+        // we make sure that the number of composite buffers will never become greater than the number of bytes in
+        // the composite buffer.
+        // This restriction guarantees that methods like countComponents, forEachReadable and forEachWritable,
+        // will never overflow their component counts.
+        // Allocating a new array unconditionally also prevents external modification of the array.
+        Buffer[] bufs = refs
+                .filter(CompositeBuffer::discardEmpty)
+                .flatMap(CompositeBuffer::flattenBuffer)
+                .toArray(Buffer[]::new);
+        // Make sure there are no duplicates among the buffers.
+        Set<Buffer> duplicatesCheck = Collections.newSetFromMap(new IdentityHashMap<>());
```

Comment:
No, we can't just remove duplicates. Composite buffers are meant to present a view of the buffers as if they had been concatenated, in order, into one. This works fine for Strings because they are immutable values and their contents gets copied, but Buffers are stateful and life-cycled, their offsets are required to line up, and the buffers live on as part of the composite.